# Urchin local adaptation
This repo was made to cleanly demonstrate how I got from raw NGS data to different sets of loci putatively under positive selection using Fst outlier, Bayenv and reduced nucleotide diversity measures.

## Step 1: From reads to bams

[Code for Step 1](https://github.com/Cpetak/urchin_adaptation/blob/main/Step1.md)

## Step 2: from bams to genotype likelihoods and PCA
[Code for Step 2](https://github.com/Cpetak/urchin_adaptation/blob/main/Step2.md)

Getting the filtered sites

[Code for Filtering steps](https://github.com/Cpetak/urchin_adaptation/blob/main/Filtering_steps.md)



## Step 3: Getting per-site Fst values - global
[Code for Step 3](https://github.com/Cpetak/urchin_adaptation/blob/main/Step3.md)


## Step 4: Running LFMM

[Code for Step 4](https://github.com/Cpetak/urchin_adaptation/blob/main/Step4.md)

## Step 5: PCAngsd

[Code for Step 5](https://github.com/Cpetak/urchin_adaptation/blob/main/Step5.md)

## Step 6: Getting per-site Fst values - pair-wise

[Code for Step 6](https://github.com/Cpetak/urchin_adaptation/blob/main/Step6.md)

## Step 7: Getting pair-wise global Fst values

<details>
  <summary>Click to view detailed code</summary>
</details>

Method 1:

```bash
./realSFS fst stats ${dir2}/pairwise_fst_cleaned/${pop1}_${pop2}_allsites.fst.idx > ${dir2}/pairwise_fst_cleaned/${pop1}_${pop2}_global.fst
```

-> output: unweighted and weighted (second preferred) global Fst for each pair. However, these are global Fst values prior to filtering based on MAF. So,

Method 2:

```python
import pandas as pd
import matplotlib.pylab as plt
import numpy as np

col_names1=["ID","chr","pos","A", "B"]

df = pd.read_csv('joined_CAP_BOD_allsites',names=col_names1, sep="\s")
df["pos"]=df.pos.astype('int64', copy=False)
#df["A"]=df.A.astype('int64', copy=False)
#df["B"]=df.B.astype('int64', copy=False)

df["fst"]=df["A"]/df["B"]
df.replace([np.inf, -np.inf], np.nan, inplace=True)
df['fst'] = df['fst'].fillna(0)

df = df.drop(["ID","A","B"], 1)

print(df["fst"].mean())
```

Results are in pairwise_global_fst.csv file. My method on the filtered dataset and the realsfs on the original dataset generated pairwise global Fst values that significantly positively correlated (p<0.001, adjusted r-squared = 0.526), and pairwise global Fst values did not correlate with distance between the populations, no matter which method was used, thus there was no isolation by distance in this WGS dataset.

Realsfs vs distance: adj r squared: 0.024, p-val = 0.238

Averaged vs distance: adj r squared: -0.053, p-val = 0.969

Insert 3 images here!

Generated by: https://colab.research.google.com/drive/10XBxj0d1lb9Wp0bcg5VX3XcXvBMolz34?usp=sharing

-> compared to bayenv cov matrix coming!

-> bootstrapped and got outliers!

Bootstrap 99th percentile cut off: 0.02607645058649987

awk -F "," ' $6 >= 0.02607645058649987 ' subbed.csv > pair_fst_outs

 awk -F "," ' $10 >= 0.23131785559999998 ' reps_combined_bayenv.csv > bayenv_outs

